src/MongoDB.Driver.Core/Core/Operations/EndTransactionOperation.cs

```diff
@@ -161,5 +161,75 @@ public CommitTransactionOperation(WriteConcern writeConcern)
         // protected properties
         /// <inheritdoc />
         protected override string CommandName => "commitTransaction";
+
+        // public methods
+        /// <inheritdoc />
+        public override BsonDocument Execute(IReadBinding binding, CancellationToken cancellationToken)
+        {
+            try
+            {
+                return base.Execute(binding, cancellationToken);
+            }
+            catch (MongoException exception) when (ShouldReplaceTransientTransactionErrorWithUnknownTransactionCommitResult(exception))
+            {
+                ReplaceTransientTransactionErrorWithUnknownTransactionCommitResult(exception);
+                throw;
+            }
+        }
+
+        /// <inheritdoc />
+        public override async Task<BsonDocument> ExecuteAsync(IReadBinding binding, CancellationToken cancellationToken)
+        {
+            try
+            {
+                return await base.ExecuteAsync(binding, cancellationToken).ConfigureAwait(false);
+            }
+            catch (MongoException exception) when (ShouldReplaceTransientTransactionErrorWithUnknownTransactionCommitResult(exception))
+            {
+                ReplaceTransientTransactionErrorWithUnknownTransactionCommitResult(exception);
+                throw;
+            }
+        }
+
+        // private methods
+        private void ReplaceTransientTransactionErrorWithUnknownTransactionCommitResult(MongoException exception)
+        {
+            exception.RemoveErrorLabel("TransientTransactionError");
+            exception.AddErrorLabel("UnknownTransactionCommitResult");
+        }
+
+        private bool ShouldReplaceTransientTransactionErrorWithUnknownTransactionCommitResult(MongoException exception)
+        {
+            if (exception is MongoConnectionException)
+            {
+                return true;
+            }
+
+            if (exception is MongoNodeIsRecoveringException)
```

Comment:
Not sure why we don't return true for MongoNotPrimaryException, but if I do the JSON tests fail.

Reply:
Which test(s) specifically?  I want to check the Java driver to see why.

Reply:
Java driver delegates this to https://github.com/mongodb/mongo-java-driver/blob/f77e7b532afe5b51b587a08ca8d46d089ceb5a52/driver-core/src/main/com/mongodb/operation/CommandOperationHelper.java#L607-L623, and all the tests pass. Let's discuss to determine why the logic is different.